### PR TITLE
Adding version requirement macro

### PR DESCRIPTION
**Pull request by ACEnglish**
New macro `todo_b4` will cause code to fail compilation if the version is Cargo.toml doesn't satisfy the requirement. e.g 

```rs
todo_b4!("<1.0.0", "Need to implement")
```

### CODE REVIEW

**Comment by parker-codes** [comment]
Hey, thanks for contributing! Here's my thoughts, and please share yours too:

1. The name `todo_b4!` is concise, but doesn't describe what is being targeted. Perhaps `todo_by_version!` could work?
2. Can you provide some use cases on the `>` or `<`  part? To me it seems like the versions would all be `>` because the version always goes up, not down.

**Comment by ACEnglish** [comment]
Hello,

I'm happy to help as this is a neat idea and was a great opportunity for me to learn a little more about rust.

1. I'm fine with whatever naming you're comfortable with. You are the designer/maintainer so whatever feels best to you is fine.
2. I originally built in a prepending of "<" to whatever `version` was provided to the macro as that is the most common use case and lines up most succinctly with `todo_by`'s `date < now`. However, I then realized manipulating the `version` string into one specific [VersionReq](https://docs.rs/semver/latest/semver/struct.VersionReq.html) was unnecessarily limiting.  A better approach would be to describe that the `todo_by_version`'s first argument is a semantic versioning requirement that's checked against the project's version in `Cargo.toml`. Whatever use a developer can come up with is possible. For example:
  * Something needs to be done before the next release `<1.0`. 
  * Something that needs to be done first after the next release `<=1.0.0`. 
  * Documenting when a todo was written and when it must be done by: `>=1.2.3, <1.8`.

**Comment by parker-codes** [comment]
Got it, thanks for clarifying. 

The range options (`>=1.2.3, <1.8`) seems a bit confusing. The `<` and `<=` make sense, but I still don't see a case for `>` because that would mean it would always fail until that version is met, which blocks a release from even happening. So if we were to implement a `todo_after!("2023-06-29")` it would have the same exact experience of showing an error once the date has passed. Does this sound right, or am I missing something?

If this is the case, then I think the control to offer is `<` vs `<=`. For dates in `todo_by!()` this is straightforward because you can just enter the previous date instead of handing the `=` case. For versioning this is more complex, obviously, because you don't know what the last minor/patch release is going to be before the next version. So I think treating them all as `<=` could work? `todo_by_version!("1.0");` would mean that once the version is 1.0 or greater, it errors.

Wildcards could be used from right-to-left by just omitting the number. So `1.0` is a wildcard patch version, and `1` is a wildcard minor and patch like `todo_by_version!("2");`

**Comment by ACEnglish** [comment]
I'm unable to understand what's confusing about the range version requirement. For `>`,  I agree it doesn't have a super reasonable use-case, but it's possible to imagine it being used as a reminder to bump the version number:
```rs
todo_by_version!(">=1.0", "While pkg version less than 1, code will fail/warn during build");
```

The larger point of not manipulating the string that's passed to [`VersionReq.parse`](https://docs.rs/semver/latest/semver/struct.VersionReq.html#method.parse) is that major/minor/patch/wildcard  and all the SemVer comparators for requirements available are handled by that lib and automatically parsed.  The macro is just piping through the todo's version requirement to a `matches -> bool` against the package's version.  

Again this is your project, so how/if you'd like to use by version is entirely up to you. For me, I tie todos to versions more frequently than dates. And for the implementation I prefer minimal intervention in the piping between the macro and semver. Sure, it allows possibly less useful requirements, but the logic seems simpler and less `todo_by` code to maintain.

**Comment by parker-codes** [comment]
Gotcha. While I don't see the use case still, I do see the benefit of not getting in the way of the semver logic. To make it more readable, I propose inverting the conditional. So `todo_by_version!(">=1.0");` means it will trigger once the version is >= to `1.0`.

**Comment by ACEnglish** [comment]
I don't know if not-negating the match is the way to go. For example, imagine we need a hotfix on some code and our solution solves the problem at hand, but needs to be properly re-implemented before getting into develop. This hotfix could be tagged with a todo that deters it from being lazily pulled in. For example:
```rs
todo_by_version!("0.2.4", "Solves this patch. Need to come up with something better");
fn works_but_awful() {}
```
Then only the 0.2.4 hotfix matches for the macro. To negate the semver requirements to work with inversion, it'd have to be something like:
```rs
todo_by_version!("<0.2.4, >0.2.4", "Solves this patch. Need to come up with something better");
fn works_but_awful() {}
```
Which I don't even know if that's a valid requirement.

**Comment by parker-codes** [comment]
I inverted it, but the `tests/failures/past_version.rs` should be failing and it isn't. I think that by not including a `>`/`<` it is defaulting to want an exact match.

I agree that having a specific version in mind breaks this logic. But it reads better to say "todo by version greater than or equal to 2.0" if it will error once it gets to `2.0`.

**Comment by parker-codes** [comment]
If the name of the macro was inverted then the previous negation would make sense, like `todo_while_version!("0.2.4")` or `todo_while_version!("<=2.0.0")`. I'm not opposed to that at all, but it would be nice to keep the macros congruent.

**Comment by ACEnglish** [comment]
Sounds good. I think the difference here is I read it as 'to do by (before) condition not met' So `todo_by!("2023-06-1")` is you have until June and `todo_by_version!("<1.0")` you have until version 1.

**Comment by parker-codes** [comment]
So are you okay with the `todo_while_version` name change? Your second example is "todo while version is less than 1.0". I guess it doesn't work well when there is no `>`/`<` though. That one is a "by".

**Comment by ACEnglish** [comment]
I guess I'm just not following. To me the second example being "todo while version is less than 1.0" could be equally applied to "todo while date (implied before/less than) 2023-06-01".

**Comment by parker-codes** [comment]
Yeah, I was confusing myself a bit. For an exact match, it still reads well: `todo_while_version!("0.5.4")` translates to "todo while version is 0.5.4". Great! 

The congruency is still an issue, but we could go back to the negation.

Another option would be to detect an exact match and handle it conditionally. It would be easiest if multiple comparators were prevented. This isn't perfect code, but what are your thoughts on this?

```rs
    if (version.comparators[0].op == semver::Op::Exact && !version.matches(&now_version))
        || version.matches(&now_version)
    {
```

**Comment by ACEnglish** [comment]
Maybe I'm being too circular in my reasoning, but this is leading me to my previous thought on how I read the macro as "todo by (before) a condition is not met`.  Or "Once this statement is false, this is no longer a todo and now a compile error".

**Comment by parker-codes** [comment]
I'm not sure where you get the "not" though.. Are you reading the macro bang (`!`) as "not"? I could infer the "before", but not a complete inversion via a "not". 😅 The date one is "todo by (before) a date is met. This one is "todo by (before) a semver condition is met".

**Comment by parker-codes** [comment]
Sadly, my solution doesn't quite fit. I scoped the comparators to just 1 and then did the matching based on `Op::Exact`, but it turns out that the default (when there are no symbols) is `Op::Caret`. Weird.

```rs
if version.comparators.len() > 1 {
    panic!("Too many semver comparators");
}
```

```rs
let is_exact_match =
    version.comparators[0].op == semver::Op::Exact && !version.matches(&current_version);
if is_exact_match || version.matches(&current_version) {
    // ...
}
```

**Comment by ACEnglish** [comment]
I get the 'not' from the original date parsing. The original date condition `date < now` - when true - raises the error. Which is equivalent to `!(now <= date)`. And the `version.matches` is equivalent to `now <= date`. Thus not condition met. The difference is that version matching allows multiple operators whereas for date you're only using `<`.

**Comment by parker-codes** [comment]
First of all, thank you for discussing this with me. I'm not trying to be a bother, but I can be picky about having a polished output. 

Your circular reasoning impressed me, haha. 😄 So the "not" you're describing is based on the internal logic, but that is hidden from the end user. They only see the API which is the. macro name.

I think if we go back to how the words work, your first macro name was `todo_b4` which reads as "do this before a version is (condition)". So to make it error once it reaches v2 or greater it seems like you could put `todo_b4!(">2.0.0")`, but it actually required `<`..

After thinking about this a lot today, "todo_by", "todo_before", etc that are putting an automatic time condition in there are misleading because the condition is already passed in via a symbol. So perhaps `todo_while_version!()` is the most straightforward because it reads as "do while the version is (condition)" which could be "do while the version is <2.0.0".

I'm going to revert back to the negation and change the macro name. Then we'll see if we like that. For me this was all about naming, and the logic change was just to try to keep it as `todo_by_version` for congruency, but that just doesn't seem to work well.

**Comment by parker-codes** [comment]
For some reason the `before_version` test is not erroring like it is supposed to.. All other tests are passing now.

File: tests/failures/before_version.rs [hunk 1]
```diff
@@ -1,6 +1,6 @@
-use todo_by::todo_by_version;
+use todo_by::todo_while_version;
 
 fn main() {
-    todo_by_version!("<0.1.0");
+    todo_while_version!("<0.1.0");
```
Comment (ACEnglish) [inline tests/failures/before_version.rs]: todo_while_version!(">0.1.0"); 

fails because the package is 0.2

**Comment by ACEnglish** [comment]
I also appreciate the conversation and think our goals of a polished output are aligned.

I think the 'before_version.rs' test might help exemplify the 'todo_by not condition' lol.

It is entirely your choice for how you'd like to name the macro. Though, I do feel that while may read the same as by, if not a little worse because some programmers may associate `while` with looping, and there's nothing looping about the macro.

I do think that it's not necessarily the case that internal logic is hidden from the user. I mean, people cannot literally see the internals, but what's being exposed is essentially `todo_by !(now <= date) `. The name of the macro helps tell the users what the left half of the conditional filled in by the macro is. Then the `date` parameter is provided by the user. Whereas for this change, it's `todo_by !(current_version matches version_requirement)`. For me, `now = current_version`, and `<=` is a subset of match operators, and  version_requirement is what's provided by the user.

I don't know if there's any operative word that captures all matches as well as closely as `by` can be inferred to mean `!(now <=` ... Maybe the documentation can make it clear that it's `todo_by` this date expires and `todo_by` this version requirement expires...  or maybe try out`todo_match`?

**Comment by ACEnglish** [comment]
Good morning. Sorry to keep flooding you with messages but this one is actually a little important....

The tl;dr is that TryBuild cannot test todo_by_version failures.

If you change the error line to output the now_version you see that when `before_version.rs` fails, it thinks the package is on version 0.0.0. I thought this was because of perhaps the new parsing or my old `nul_version`, but even this parsing (sans nul_version) gave 0.0.0

```rs
let now_version = match Manifest::from_path("Cargo.toml") {
        Ok(a) => match a.package {
            Some(p) => match p.version.get() {
                Ok(v) => Version::parse(v).unwrap(),
                Err(_) => panic!("Cannot happen"),
            },
            None => panic!("Cannot happen"),
        },
        Err(_) => panic!("Cannot happen"),
    };
```

I suspect TryBuild because when you change a basic.rs test to check e.g. <0.1.0, or you manually change the `Cargo.toml` to `version = 100.2.1`, the basic.rs tests fail and print the real Cargo.toml version (0.2.1 or 100.2.1 respectively).

Digging in more, I saw that trybuild talks about how they run rustc on the failures, and this command:

```
cargo test -vv -- todo_by trybuild=tests/failures/before_version.rs
```
Shows that todo_by major/minor versions are passed to the command via environment variables, but manually logging the Cargo.toml path shows that TryBuild is making its own 0.0.0 version inside `target/tests/trybuild/todo_by/Cargo.toml`. (uncomment lines 84-86 to reproduce).

**Comment by parker-codes** [comment]
This makes sense! So there may be a way for us to stage trybuild-specific tests at compile-time using `#[cfg(trybuild)]`. 

I found this out in https://github.com/parker-codes/todo_by/pull/5 where I'm working on finding the best DX for the error/warning system.

Maybe if it's running in trybuild, we pull from a `TRYBUILD_STUB_VERSION` envar or something, and then we'd just need to set that value in each test.

**Comment by ACEnglish** [comment]
I'll have to look into that more to think about it. Maybe I can help outside of this little version thing.

One last note on version, I just realized that my  last commit's more informative error may have solved the by/while/before question..

```TODO version requirement '{version_str}' not satisfied by {current_version}```
Which is the same structure as
```TODO date requirement '{date}' not satisfied by {current_date}```

Both are `TODO (user arg) by now`.. Also, `<2023-05-01` and  `<0.1.0` both fail for the current date/version . Difference is that date has an implied less than... 
Okay I'm done  😄

**Comment by parker-codes** [comment]
You're right about the trybuild test failures. I read through the source code and they hardcode the version to `0.0.0`. So we could consider suggesting a patch, but there may be a better way.

I attempted the envar route I mentioned above (setting something like `TRYBUILD_STUB_VERSION` and then conditionally accessing it in `current_version_str`, but it doesn't work because the `todo_while_version` macro expansion happens at compile time, while the envar being set is at runtime. We could perhaps set that before running `cargo test`, but then failures might need to be run individually. I don't love that idea.. I believe I found the solution, and it's pretty clean. We can use this strategy and set the envar in the `failures` test but test them individually. I'm working on this now.

**Comment by parker-codes** [comment]
Turns out my strategy works, but `#[cfg(trybuild)]` is not actually working properly. But this direction gave me an idea, and may solve the issue where tests break because of these macros.

**Comment by ACEnglish** [comment]
I was thinking about this a little yesterday and came to a realization. First, I want to make it very clear that I'm very pro-tests, so please don't interpret anything I'm saying as trying to skirt test coverage responsibilities. 

I just put the `before_version.rs` test back in. It fails appropriately with `todo_by_version!(">0.1.0");` because `now` (0.0.0 from trybuild) is before the requirement. We were trying to test `now` is after failure (e.g. 2.0 not satisfying <1.0), but there's no before 0.0.0. But I think that's okay because todo_by is covered test-wise because we touch every line of code.  At some point I think it's fine to trust semver and [their testing](https://github.com/dtolnay/semver/blob/master/tests/test_version_req.rs) to cover all the comparison functionality.

**Comment by parker-codes** [comment]
I'm going to kill 2 birds with one stone here.

**Comment by parker-codes** [comment]
@ACEnglish Now all the tests are passing with stubbed values! I still want to eventually disable actual test failures, but I'll leave that for another PR.

I did run across a different problem though - a regression:

Screenshot 2023-05-17 at 6 18 14 PM

Perhaps the previous code nested the block better, so I'll try that.

**Comment by parker-codes** [comment]
I think this is in a good spot to merge. I really do want to find a better strategy for disabling during tests too, while still keeping the error message.

**Comment by ACEnglish** [comment]
Sounds good. I have renamed it to `todo_while` since it's not `todo_by_date` and it felt clunky to to have one of the  macros three times as long as the other.

Nice working with  you on this!
